src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ *  Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ *  This code is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 only, as
+ *  published by the Free Software Foundation.  Oracle designates this
+ *  particular file as subject to the "Classpath" exception as provided
+ *  by Oracle in the LICENSE file that accompanied this code.
+ *
+ *  This code is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package jdk.internal.foreign;
+
+import jdk.incubator.foreign.MemorySegment;
+import jdk.internal.access.JavaNioAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * Implementation for heap memory segments. An heap memory segment is composed by an offset and
+ * a base object (typically an array). To enhance performances, the access to the base object needs to feature
+ * sharp type information, as well as sharp null-check information. For this reason, the factories for heap segments
+ * use a lambda to implement the base object accessor, so that the type information will remain sharp (e.g.
+ * the static compiler will generate specialized base accessor for us).
+ */
+public class HeapMemorySegmentImpl<H> extends AbstractMemorySegmentImpl {
+
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+    private static final int BYTE_ARR_BASE = UNSAFE.arrayBaseOffset(byte[].class);
+
+    final long offset;
+    final Supplier<H> baseProvider;
+
+    @ForceInline
+    HeapMemorySegmentImpl(long offset, Supplier<H> baseProvider, long length, int mask, Thread owner, MemoryScope scope) {
```

Comment:
Just tried this - makes benchmarks revert to old behavior - roughly 10x slower. Worth a try, but I think using reflection probably makes thing less transparent to the JIT.

Reply:
Interesting, I thought it might be the explicit casts introduced in the MH machinery of lambdas, and since everything is final and stable (?) the casts usually help the JIT, if so then there is clearly more going on with the MH machinery.

Reply:
that's true, although I'd expect that with MH the Class constant is stored in some separate spinned class - e.g. it is a true constant, not a final field load?

Reply:
Maybe the speedup comes from each lambda having a separate type (a class is spun for each), so C2 can apply type profiling & guarded inlining, and within the inlined section a sharper array type can be assumed.